cfep-13.md

```diff
@@ -0,0 +1,70 @@
+
+<table>
+<tr><td> Title </td><td> Securing `conda-forge` Uploads to `anaconda.org` </td>
+<tr><td> Status </td><td> Proposed </td></tr>
+<tr><td> Author(s) </td><td> Matt, Marius, Anthony, Isuru</td></tr>
+<tr><td> Created </td><td> Feb 27, 2020</td></tr>
+<tr><td> Updated </td><td> Feb 27, 2020</td></tr>
+<tr><td> Discussion </td><td> NA </td></tr>
+<tr><td> Implementation </td><td> NA </td></tr>
+</table>
+
+## Abstract
+
+Currently the `anaconda.org` tokens on each feedstock allow any `conda-forge`
+feedstock maintainer to upload any output to any `anaconda.org` package.
+This could allow someone to push a compromised and/or malicious build
+of say the `python` package. This situation is a security risk that we
+should mitigate.
+
+## Specification
+
+We will secure the uploads of outputs to `anaconda.org` via the following process.
+
+1. Each feedstock is provisioned with a secret key that is only known to the
+   maintainers of that feedstock and the core team. These keys will be kept in
+   a private repo with only core members having read access. Core members and 
+   encrypted GitHub keys for bot admins will have write access in order to generate 
+   the secret keys and push then to the CI providers.
+2. A public, global registry of all outputs from all feedstocks will be kept
+   in a repo on github. Write access to this repo is limited to core and bot 
+   admins as well.
+3. When a feedstock builds new outputs it wants to upload, it will push these outputs
+   to a staging organization on `anaconda.org`. It will also send the names of the
+   outputs, their SHA256 checksums, the feedstock name, and the secret key to a
```

Comment:
Are we literally sending the secret key? Better to use public/private keys and sign the message to the web service, right?

Reply:
A lot of services send API keys over https these days (in the authorization bearer headers), so I figured that would be fine here. Are you saying that is not secure enough for this purpose?

Reply:
I mean, I suppose https is encrypted anyway, and I'm definitely no expert on these things. It just feels more potentially risky in case of accidentally sending requests to the wrong place / etc.

Reply:
Hmmm. At a practical level, either an API key or the private key (to sign messages) needs to be stored on the CI services. For that point of view the risk for spilling it is mostly the same and due to the CI service itself, maintainers making edits etc IIUIC. 

I am not personally familiar with how to store a private key on the CI services or if they are setup to handle such things. Should we put it in an env var?

Unless we have a good answer for how to store a private key, I’m inclined to go the API key route since I don’t think I will mess it up as easily.

Reply:
From reading the text, it is not clear to me what the secret key is used for. Is this for signing? Then using async signing should allow us to have a public key for that?

Reply:
The secret key is a standard ApI key. 

Do you know how to put private keys in the CI services?

If we embed an entire file in an env var, will the CO service properly mask it in outputs?

Reply:
You can store private keys in CI services, depends a bit on the service. Travis provides a different public encryption key per repo, for Github Actions you can set it via the UI in Actions Tab (and there is hopefully an API for that), not sure about Azure but I would expect to provide a similar thing. They all will mask the exact value which is in the key in the CI log. The key will also be provided as an env var in the CI job but only when run as part of a commit in `conda-forge`, not in PRs.

Reply:
This entire issue around an API key versus public/private keys is a red herring. 

My main concern is correctly implementing this system both in the code and as a whole with various keys secured correctly. 

I’m assuming I’ll be doing the coding here and I am not confident I can handle the public/private key option correctly. 

A worse outcome is for us to think we have fixed this when there is actually a bug. This is especially true with respect to cryptography. 

If someone else wants to help code this with public/private keys and knows definitely how to secure the private key on our CI services, then I am happy to do that route. 

Otherwise let’s use an API key sent over https since that can be done by me confidently.

Reply:
Setting keys via a UI is a nonstarter. 

It’s not the env var I’m worried about @xhochy. Do you know for sure that an env var with new lines will be masked properly in all of our CI services? Usually people stuff API keys in these, not full files.

Reply:
> It’s not the env var I’m worried about @xhochy. Do you know for sure that an env var with new lines will be masked properly in all of our CI services? Usually people stuff API keys in these, not full files.

With full files, I have no experience, so: no.

Reply:
If the worry is about multi-line text, it could probably be base64-encoded for env var storage, no?

Reply:
Unfortunately I absolutely do not have the spare cycles to help in any concrete way :-( For completeness, [base64](https://en.wikipedia.org/wiki/Base64) encodings can help pack arbitrary binary data into things like environment variables.

Reply:
cool - let's go with the API key approach for now. If you get spare time and have interest, then feel free to pitch in!

Reply:
`base64` might not be the best choice if the CI would decide to only mask the encoded but not the decoded key :laughing:.
How ever you want to go with it, this would also work:
```sh
# put it in one line:
key_as_one_line="$( < key-file head -n-1 | tail -n+2 | tr -d \\n )"
# reassemble and use:
echo -e "-----BEGIN PRIVATE KEY-----\n${key_as_one_line}\n-----END PRIVATE KEY-----" | \
  openssl dgst -sha256 -sign - some-file > some-file.sha256
```